### PR TITLE
Drop sessions with duplicate UE IP

### PR DESCRIPTION
**Pull request by ivan4th**
Drop the conflicting session upon a UE IP conflict, sending
Session Report Request for the dropped session as per
TS 29.244 clause 5.18.2: UP Function Initiated PFCP Session Release

### CODE REVIEW

File: upf/upf_pfcp_api.c [hunk 1]
```diff
@@ -657,28 +657,60 @@ handle_f_teid (upf_session_t * sx, upf_main_t * gtm, pfcp_pdi_t * pdi,
...
   upf_main_t *gtm = &upf_main;
   const dpo_id_t *dpo;
+  int r = 0;
 
   if (ue_addr->flags & IE_UE_IP_ADDRESS_V4)
     {
       dpo = upf_get_session_dpo_ip4 (nwi, &ue_addr->ip4);
       if (dpo && dpo->dpoi_index != sx - gtm->sessions)
-	return -1;
+	{
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:695]: a real problem only exists if the same CP node tries to use the same UE IP twice. If the UE IP was allocated by a different node, that other node would still be aware of the assignment and there would be no further problems.

The function should therefore check for the node_id and only remove the other session if the node_ids match.

File: upf/upf_pfcp_api.c [hunk 2]
```diff
@@ -2505,7 +2551,7 @@ handle_session_modification_request (pfcp_msg_t * msg,
   if (r == 0)
     resp->cause = PFCP_CAUSE_REQUEST_ACCEPTED;
 
-  upf_pfcp_send_response (msg, &resp_dmsg);
+  upf_pfcp_send_response (msg, &resp_dmsg, outlives_session);
```
Comment (RoadRunnr) [inline upf/upf_pfcp_api.c:2554]: there is on thing that bugs me about the `outlives_session` argument, why do we keep a reference between session and messages in the response cache and request queue anyway?
There should be no reason to remove a message from the response cache even when the session is deleted. A possible resent of the response does not need to access the session in any way.
The request queue is a bit more complex,
any answer should remove the message from the queue, but it should not attempt to access the session if the session is already gone.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: upf/upf_pfcp_server.c [hunk 3]
```diff
@@ -517,6 +517,53 @@ urr_check_counter (u64 bytes, u64 consumed, u64 threshold, u64 quota)
...
+
+  SET_BIT (req->grp.fields, SESSION_REPORT_REQUEST_PFCPSRREQ_FLAGS);
+  /* PSDBU = PFCP Session Deleted By the UP function */
+  req->pfcpsrreq_flags = PFCPSRREQ_PSDBU;
+
+  /*
+   * outlives_session=true so that this request isn't dropped
+   * from the response queue after the session is deleted,
+   * and can be retransmitted if it's lost
+   */
```
Comment (RoadRunnr) [inline upf/upf_pfcp_server.c:562]: that comment is obsolete now.

File: upf/upf_pfcp_api.c [hunk 4]
```diff
@@ -657,28 +657,60 @@ handle_f_teid (upf_session_t * sx, upf_main_t * gtm, pfcp_pdi_t * pdi,
 
 }
 
+static void
+expunge_conflicting_session (upf_session_t * sx_old,
```
[the remark this answers is not in the document]
Comment (ivan4th) [reply upf/upf_pfcp_api.c]: Fixed

File: upf/upf_pfcp_api.c (hunk 1 above)
Comment (ivan4th) [reply upf/upf_pfcp_api.c:698]: Fixed. If node id is different, the new Session Establishment Request will still fail, but the old session will not be dropped.

File: upf/upf_pfcp_api.c (hunk 2 above)
Comment (ivan4th) [reply upf/upf_pfcp_api.c]: Removed the code that was dropping session-related messages upon session deletion, and updated the tests accordingly.

File: upf/upf_pfcp_server.c (hunk 3 above)
Comment (ivan4th) [reply upf/upf_pfcp_server.c]: Removed.

**Comment by ivan4th** [comment]
Had to include some not-completely-related e2e changes that I accumulated (PFCP request pipelining) so as to avoid serious merge conflicts in the E2E tests later.

**Review by sergeymatov — APPROVED** [review]

**Review by RoadRunnr — APPROVED** [review]

**Pull request merged**